src/coins.cpp

```diff
@@ -265,22 +283,84 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
             it++;
     }
     hashBlock = hashBlockIn;
+    if (nBestCoinHeightIn > nBestCoinHeight)
+        nBestCoinHeight = nBestCoinHeightIn;
+
     return true;
 }
 
 bool CCoinsViewCache::Flush()
 {
     LOCK(cs_utxo);
-    bool fOk = base->BatchWrite(cacheCoins, hashBlock, cachedCoinsUsage);
+    bool fOk = base->BatchWrite(cacheCoins, hashBlock, nBestCoinHeight, cachedCoinsUsage);
     return fOk;
 }
 
 void CCoinsViewCache::Trim(size_t nTrimSize) const
 {
-    uint64_t nTrimmed = 0;
-
     LOCK(cs_utxo);
+
+    uint64_t nTrimmed = 0;
+    uint64_t nTrimmedByHeight = 0;
+    static uint64_t nTrimHeightDelta = nBestCoinHeight * 0.80; // This is where we attempt to do our first trim
+    int64_t nTrimHeight = nBestCoinHeight - nTrimHeightDelta;
+
+    // Begin first Trim loop. This loop will trim coins from cache by the coin height, removing the oldest coins first.
+    // This has been proven to improve sync performance significantly for nodes that can not hold the entire dbcache
+    // in memory.
+    bool fDone = false;
+    uint64_t nSmallestDelta = 50; // number of blocks to adjust trim height by
     CCoinsMap::iterator iter = cacheCoins.begin();
+    while (!fDone && DynamicMemoryUsage() > nTrimSize)
+    {
+        LogPrint("coindb", "cacheCoinsUsage at start: %d total dynamic usage: %d trim to size: %d nBestCoinHeight: %d "
+                           "trim height:%d\n",
+            cachedCoinsUsage, DynamicMemoryUsage(), nTrimSize, nBestCoinHeight, nTrimHeight);
+
+        iter = cacheCoins.begin();
+        while (DynamicMemoryUsage() > nTrimSize)
+        {
+            if (iter == cacheCoins.end())
+            {
+                fDone = true;
+                break;
+            }
+
+            if (iter->second.flags == 0 && iter->second.coin.nHeight < nTrimHeight)
+            {
+                cachedCoinsUsage -= iter->second.coin.DynamicMemoryUsage();
+
+                iter = cacheCoins.erase(iter);
+                nTrimmed++;
+                nTrimmedByHeight++;
+            }
+            else
+                iter++;
+        }
+
+        // Gradually increase the nTrimHeight if we didn't trim enought entries.
+        if (fDone && DynamicMemoryUsage() > nTrimSize && nTrimHeightDelta > nSmallestDelta)
+        {
+            if (nTrimHeightDelta <= nSmallestDelta * 100)
+                nTrimHeightDelta -= (nSmallestDelta * 2);
+            else if (nTrimHeightDelta <= nSmallestDelta * 400)
+                nTrimHeightDelta -= (nSmallestDelta * 10);
+            else
+                nTrimHeightDelta -= (nSmallestDelta * 200);
+
+            nTrimHeight = nBestCoinHeight - nTrimHeightDelta;
+
```

Comment:
Use saturating arithmetic: I'd like to see logic that ensures that nTrimHeightDelta and nTrimHeight stop at zero rather than wrap around and cause weird behavior.

Reply:
ok , done...